pysparkling/cache_manager.py

```diff
@@ -47,14 +47,14 @@ def add(self, ident, obj, storageLevel=None):
             'disk_location': None,
             'checksum': None,
         }
-        log.debug('Added {0} to cache.'.format(ident))
+        log.debug(f'Added {ident} to cache.')
```

Comment:
This wasn't great before, but now that pylint complained about it, this should be fixed: I'd vote for using the native formatter of the logger. In that way, when logging is below the level, no new string is generated.

For example: even when nothing is logged, the above code would make a string and then do nothing with it. However, `log.debug('Added %s to cache', ident)` is a no-op.

Reply:
Ok, I was doubting what would be the best, and I'll adjust it based on your comment.

Reply:
Actually, I remember my train of thought at the time:
So we have 3 possibilities:

```python
import logging
import time

logging.basicConfig()


class HeavyOperation:
    @property
    def id(self):
        time.sleep(0.5)
        return 123

def _try(log_level, n=10):
    log.setLevel(log_level)

    start = time.time()
    for _ in range(n):
        log.debug('Added {0} to cache.'.format(ho.id))
    print(f'time to log (format) under {log_level}: {(time.time() - start)/n:.5f}')

    start = time.time()
    for _ in range(n):
        log.debug('Added %s to cache.', ho.id)
    print(f'time to log (lazy) under {log_level}: {(time.time() - start)/n:.5f}')

    start = time.time()
    for _ in range(n):
        log.debug(f'Added {ho.id} to cache.')
    print(f'time to log (f-string) under {log_level}: {(time.time() - start)/n:.5f}')

log = logging.getLogger(__name__)

ho = HeavyOperation()

_try('DEBUG')
_try('INFO')
```

Output:
```
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
time to log (format) under DEBUG: 0.50043
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
time to log (lazy) under DEBUG: 0.50062
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
DEBUG:__main__:Added 123 to cache.
time to log (f-string) under DEBUG: 0.50058
time to log (format) under INFO: 0.50346
time to log (lazy) under INFO: 0.50049
time to log (f-string) under INFO: 0.50053

Process finished with exit code 0
```

What I want to say is that there is a negligible difference between the cases.
The main reason I see is because of the interpretation BEFORE passing it into the logging method.

If however there is no processing needed (unlike my example), then I don't even think it's worth writing it as %-strings.

I prefer readability of the code at that point.

For example:

```python
log.debug('Returning %s from cache.', ident)

log.debug(f'Returning {ident} from cache.')
```

I consider the 2nd the cleanest.

```python
import logging
import time

# logging.basicConfig()  # --> Don't output


def _try(log_level, n=100_000):
    log.setLevel(log_level)

    ident = 123

    start = time.time()
    for _ in range(n):
        log.debug('Returning %s from cache.', ident)
    print(f'time to log (lazy) under {log_level}: {(time.time() - start):.5f}')

    start = time.time()
    for _ in range(n):
        log.debug('Returning {0} from cache.'.format(ident))
    print(f'time to log (format) under {log_level}: {(time.time() - start):.5f}')

    start = time.time()
    for _ in range(n):
        log.debug(f'Returning {ident} from cache.')
    print(f'time to log (f-string) under {log_level}: {(time.time() - start):.5f}')


log = logging.getLogger(__name__)

_try('DEBUG')
_try('INFO')
```

Output:
```
time to log (lazy) under DEBUG: 1.12379
time to log (format) under DEBUG: 0.93306
time to log (f-string) under DEBUG: 0.97145
time to log (lazy) under INFO: 0.02394
time to log (format) under INFO: 0.04687
time to log (f-string) under INFO: 0.03790
```

Mind you, these are total timings for 100k loops!

I would still prefer (at least for the cases where it is nicer) to go with the f-string implementation as I don't think any of the parameters in the logging would cause excessive computations?

Reply:
```python
        log.warning(
            f'Attempt {task_context.attempt_number} failed '
            f'for partition {partition.index} '
            f'of {rdd.name()} (id: {rdd.id()}): '
            f'{traceback.format_exc()}'
        )
```
vs
```python
        log.warning(
            'Attempt %s failed for partition %s of %s (id: %s): %s',
            task_context.attempt_number,
            partition.index,
            rdd.name(),
            rdd.id(),
            traceback.format_exc()
        )
```

I would prefer the fstring one because it's immediately obvious what parameter is belong at what place. For the lazy logging, I would need to use my brain if for example I want to add a parameter in the middle. Otherwise, I just add it.

Simplicity trumps a 0,0000001396 delay per parameter ;-)

Reply:
Thanks for this careful analysis. Applications might be very different and this readability in log messages is not worth it. Let's stay with the native style.

Reply:
Readability is all in code (at least in maintainable code).

In another comment you mentioned to reduce the indirections, but here you want me to actually increase them (or at the very least, I reduced them by using fstrings, and you want me to revert this change)... As I see it: you got to lookup what parameter belongs at what position, vs reading a normal sentence.

I could understand if the main concern is about performance, but nearly all of them are just integers or strings being combined into a log message. And yes, they might not even be used, but no-one will even notice this "hit" on the performance.

So I strongly recommend to keep the f-string way here, even for debug messages. It is so much more natural when reading through the code!

==> I'm going to ask around at my working place as there are a lot of knowledgeable people there which might provide some additional insights.

Reply:
- https://github.com/PyCQA/pylint/issues/2395

- https://docs.python.org/3/howto/logging.html#optimization
- https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/ << Aggregation of log messages might be important!

Important remark: https://github.com/PyCQA/pylint/issues/2395#issuecomment-511992751
> 1. if interpolation fails, it will be logged instead of raising an exception
> 2. things like Sentry can aggregate log messages. If you interpolate before logging, Sentry would see it as one event 0

https://github.com/PyCQA/pylint/issues/2395#issuecomment-544697757
> The run-time implications and side-effects of the f-string eval() simply don't belong in logging statements.

https://github.com/PyCQA/pylint/issues/2395#issuecomment-546004944
> it's generally best practice and more pythonic to explicitly catch your exceptions. Trusting the logger to implicitly catch exceptions, but only when the code is contained within the logging statement, seems like a rather odd and unreliable approach.


I would very much value your opinion about this though to increase my understanding of the language!

Reply:
BTW, I just noticed that a lot of the logging messages were using .format() in the past, so in that sense there is no difference with the performance of an f-string.

Reply:
Yes, after doing my research I was a bit on the fence about the potential benefits/drawbacks. However I would still like to know your way of thinking about why you consider the lazy style more correct? (% - style is ambiguous because it could also mean this way of formatting: `"test %d" % 123` ;-).